File: 普通数组/56.合并区间.java
// merge 合并重叠区间，返回一个不重叠且覆盖所有输入区间的区间数组。
// 思路：先按区间左端点排序，然后线性扫描，维护当前待合并区间 cur。
// 遇到下一个区间若与 cur 重叠（next.start <= cur.end），则更新 cur.end = max(cur.end, next.end)；
// 否则将 cur 入结果并把 cur 更新为下一个区间。
// 时间复杂度：O(n log n)（排序），空间复杂度：O(n)（用于结果）。
class Solution {
    public int[][] merge(int[][] intervals) {
        if(intervals==null || intervals.length == 0){
            return new int[0][];
        }
        Arrays.sort(intervals,(a,b)->{
            if(a[0] == b[0]){
                return Integer.compare(a[1],b[1]);
            }
            return Integer.compare(a[0],b[0]);
        });
        List<int[]> list = new ArrayList<>();
        int[] res = new int[2];
        res[0]=intervals[0][0];
        res[1]=intervals[0][1];
        for(int i=1;i<intervals.length;i++){
            if(intervals[i][0]<=res[1]){
                res[1]=Math.max(intervals[i][1],res[1]);
            }else{
                list.add(res);
                res=new int[2];
                res[0]=intervals[i][0];
                res[1]=intervals[i][1];
            }
        }
        list.add(res);
        return list.toArray(new int[list.size()][]);
    }
}

File: 普通数组/56.合并区间.go
package 普通数组
func merge(intervals [][]int) [][]int {
    n:=len(intervals)
    if n==0 {
        return nil
    }
    sort.Slice(intervals,func(i,j int)bool{
        if intervals[i][0] == intervals[j][0]{
            return intervals[i][1] < intervals[j][1]
        }
        return intervals[i][0]<intervals[j][0]
    })

    res:=make([][]int,0,n)
    cur := make([]int,2)
    copy(cur,intervals[0])
    for i:=1;i<n;i++{
        next:=intervals[i]
        if next[0]<=cur[1]{
            if next[1]>cur[1]{
                cur[1]=next[1]
            }
        }else{
            res = append(res,cur)
            cur = make([]int,2)
            copy(cur,next)
        }
    }
    res = append(res,cur)
    return res
}